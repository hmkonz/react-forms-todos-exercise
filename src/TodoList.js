import React, {useState} from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { v4 as uuid} from 'uuid';

const TodoList = () => {
    // set state (todos) initially to an empty array
    const [todos, setTodos] = useState([]);
    // 'addTodo' is a function that's passed down to 'form' which is called when submit form. Adds a new todo to 'todos' state
    const addTodo = ({todoName}) => {
        // reset state (todos) by creating a new array with all the existing todos in 'todos' as well as a new object with the id and form input value of todoName of new todo
        setTodos(todos => [...todos, {id:uuid(), todoName}])
    }
  
    // 'id' passed in to remove() is the id of the todo to be removed
    const remove = id => { 
        // filter() interates over the in-state 'todos' array and passes in each id through the callback function. When the id of the todo being looped over todo.id !== id (id of todo to be deleted) returns false (means todo.id===id), that 'todo' element from the array is not added to the new array created. SetTodos is now passed that new array and changes the stored state of 'todos' to this new array without the todo that was deleted.
            setTodos(todos => todos.filter(todo => todo.id !== id));
        }
    
    return (
            <div>
                {/* render NewTodoForm with the addTodo function passed in so addTodo can be called when form is submitted */}
                <NewTodoForm addTodo={addTodo} />
                {/* iterate over each todo in the todos array and render the Todo component with the values of what's in state in 'formData' passed in  */}
                <ul>{todos.map(todo => ( 
                    <Todo 
                        key={todo.id}
                        id={todo.id} 
                        todoName={todo.todoName}
                        handleRemove={remove} 
                    />
                    ))} 
                </ul> 
            </div>    
        
        )
    }
    
    export default TodoList;
 

