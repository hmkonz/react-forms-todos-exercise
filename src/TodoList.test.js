import '@testing-library/jest-dom/extend-expect';
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import TodoList from "./TodoList";

// smoke test
it("renders without crashing", function() {
    render (<TodoList />);
});

// snapshot test
it("matches snapshot", function() {
    const {asFragment} = render (<TodoList />);
    expect (asFragment()).toMatchSnapshot();
});


it("can add a new todo", function() {
    const {getByLabelText, queryByText, getByPlaceholderText} = render (<TodoList />);
    
    // get the input value of Task as well as the button from the form
    const todoNameInput = getByLabelText("Task:");
    const button = queryByText("Add a Task");
    
    // no todos added yet so confirm there isn't a todo and a button with an "X" which is added with every todo
    expect (queryByText("Do Laundry")).not.toBeInTheDocument();
    expect(queryByText("X")).not.toBeInTheDocument();

    // Add a todo (by mock filling out the form)
    // Mocks what will be filled out in the todoNameInput input of the form and updates the state of "todos"
    fireEvent.change(todoNameInput, {target: {value: "Do Laundry"}});
    //mocks submitting the button 
    fireEvent.click(button);

    // when todo is added, expect to see a todo with a button with an 'X" on it
    expect(queryByText("Do Laundry")).toBeInTheDocument();
    expect(queryByText("X")).toBeInTheDocument();

    // after todo is added, inputs are cleared and expect to see placeholders in inputs
    const todoNamePlaceholder = getByPlaceholderText("Enter a task");
    expect(todoNamePlaceholder).toBeInTheDocument();
});

it("can delete a todo", function() {
    const {queryByText, getByLabelText} = render (<TodoList />)

    // get the input value of Task as well as the button from the form
    const todoNameInput = getByLabelText("Task:");
    const button = queryByText("Add a Task");

    // add a todo (by mock filling out the form)
    fireEvent.change(todoNameInput, {target: {value: "Do Laundry"}});
    //mocks submitting the button 
    fireEvent.click(button);
    
    // click the remove button ('X') and the todo should be gone
    fireEvent.click(queryByText("X"));
    expect(queryByText("X")).not.toBeInTheDocument();
})

