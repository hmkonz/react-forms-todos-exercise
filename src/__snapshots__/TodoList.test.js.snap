// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="todoName"
      >
        Task:
      </label>
      <input
        id="todoName"
        name="todoName"
        placeholder="Enter a task"
        type="text"
        value=""
      />
      <button>
        Add a Task
      </button>
    </form>
    <ul />
  </div>
</DocumentFragment>
`;
